routes/spotifyRouter.js

```diff
@@ -23,11 +39,27 @@ router.get('/login', (req, res) => {
     let spotifyApi = req.app.locals.main.spotifyApi
     let html = spotifyApi.createAuthorizeURL(scopes);
     console.log(html);
-    res.redirect(html + "&show_dialog=true")
+
+    let state = generateRandomString(16);
+    res.cookie(stateKey, state);
+
+    let qs = require('querystring');
+    res.redirect(html + "&state=" + state + "&show_dialog=true")
 });
 
 router.get('/callback', async (req, res) => {
-    // TODO: callback not safe, need to use implement random hash string to encrypt callback, look at Spotify docs
+    let state = req.query.state || null;
+    let storedState = req.cookies ? req.cookies[stateKey] : null;
+
+    if (state === null || state !== storedState) {
+        let qs = require('querystring');
+        res.redirect('/#' +
+            qs.stringify({
+                error: 'state_mismatch'
+            }));
+        return;
+    }
+
```

Comment:
I'm not exactly 100% too sure why this is needed, but basically, let's say someone intercepts the callback from another request, then they may access the user's access token and refresh token and mess up the user's Spotify profile. 

But by storing our state in our original request, we ensure that no one else's request can get the original requester's tokens.

Gotten from: https://github.com/spotify/web-api-auth-examples/blob/master/authorization_code/app.js

Reply:
Hmm I see!! Do you remember where you found the info/explanation for this?

Reply:
Unfortunately no ): but I think once we have almost everything polished up and ready we need to look into security